File: support/handmade/.json/jsonReader.go
// 以下是模板驗證用程式碼

package _json

import (
	"encoding/json"
	"fmt"
)

type RewardReader struct {
	*RewardStorer
}

func (this *RewardReader) FileName() FileName {
	return NewFileName("reward", ".json")
}

func (this *RewardReader) FromData(data []byte) error {
	this.RewardStorer = &RewardStorer{
		Datas: map[int64]*Reward{},
	}

	if err := json.Unmarshal(data, this.RewardStorer); err != nil {
		return fmt.Errorf("from data failed: %w", err)
	}

	return nil
}

func (this *RewardReader) MergeData(data []byte) error {
	tmpl := &RewardStorer{
		Datas: map[int64]*Reward{},
	}

	if err := json.Unmarshal(data, tmpl); err != nil {
		return fmt.Errorf("merge data failed: %w", err)
	}

	if this.RewardStorer == nil {
		this.RewardStorer = &RewardStorer{
			Datas: map[int64]*Reward{},
		}
	}

	for k, v := range tmpl.Datas {
		if _, ok := this.RewardStorer.Datas[k]; ok {
			return fmt.Errorf("merge data failed: key repeat")
		}

		this.RewardStorer.Datas[k] = v
	}

	return nil
}

func (this *RewardReader) Clear() {
	this.RewardStorer = nil
}

func (this *RewardReader) Get(key int64) (result *Reward, ok bool) {
	result, ok = this.Datas[key]
	return result, ok
}

func (this *RewardReader) Keys() (result []int64) {
	for itor := range this.Datas {
		result = append(result, itor)
	}

	return result
}

func (this *RewardReader) Values() (result []*Reward) {
	for _, itor := range this.Datas {
		result = append(result, itor)
	}

	return result
}

func (this *RewardReader) Count() int {
	return len(this.Datas)
}

// 以下是為了通過編譯的程式碼, 不可使用

File: support/handmade/.json/jsonReader.cs
// 以下是模板驗證用程式碼
// using區段可能與實際給的不一致, 要注意

using Newtonsoft.Json;
using System.Collections.Generic;

namespace SheeterJson {
    using Data_ = Reward;
    using PKey_ = System.Int64;
    using Storer_ = RewardStorer;

    public partial class RewardReader : Reader {
        public string DataName() {
            return "reward";
        }

        public string DataExt() {
            return "json";
        }

        public string DataFile() {
            return "reward.json";
        }

        public string FromData(string data) {
            Storer_ result;

            try {
                result = JsonConvert.DeserializeObject<Storer_>(data);
            } catch {
                return "from data failed: deserialize failed";
            }

            if (result == null)
                return "from data failed: result null";

            storer = result;
            return string.Empty;
        }

        public string MergeData(string data) {
            Storer_ result;

            try {
                result = JsonConvert.DeserializeObject<Storer_>(data);
            } catch {
                return "merge data failed: deserialize failed";
            }

            if (result == null)
                return "merge data failed: result null";

            foreach (var itor in result.Datas) {
                if (storer.Datas.ContainsKey(itor.Key))
                    return "merge data failed: key repeat";

                storer.Datas[itor.Key] = itor.Value;
            }

            return string.Empty;
        }

        public void Clear() {
            storer.Datas.Clear();
        }

        public bool TryGetValue(PKey_ key, out Data_ value) {
            return storer.Datas.TryGetValue(key, out value);
        }

        public bool ContainsKey(PKey_ key) {
            return storer.Datas.ContainsKey(key);
        }

        public IEnumerator<KeyValuePair<PKey_, Data_>> GetEnumerator() {
            return storer.Datas.GetEnumerator();
        }

        public Data_ this[PKey_ key] {
            get {
                return storer.Datas[key];
            }
        }

        public ICollection<PKey_> Keys {
            get {
                return storer.Datas.Keys;
            }
        }

        public ICollection<Data_> Values {
            get {
                return storer.Datas.Values;
            }
        }

        public int Count {
            get {
                return storer.Datas.Count;
            }
        }

        private Storer_ storer = new Storer_();
    }
}

// 以下是為了通過編譯的程式碼, 不可使用